src/uct/tcp/tcp_cm.c

```diff
@@ -8,82 +8,266 @@
 #include <ucs/async/async.h>
 
 
-unsigned uct_tcp_cm_conn_progress(uct_tcp_ep_t *ep)
+void uct_tcp_cm_change_conn_state(uct_tcp_ep_t *ep, uct_tcp_ep_conn_state_t new_conn_state)
 {
     uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
                                             uct_tcp_iface_t);
+    char str_local_addr[UCS_SOCKADDR_STRING_LEN];
+    char str_remote_addr[UCS_SOCKADDR_STRING_LEN];
+    uct_tcp_ep_conn_state_t old_conn_state;
+
+    old_conn_state = ep->conn_state;
+    ep->conn_state = new_conn_state;
+
+    switch(ep->conn_state) {
+    case UCT_TCP_EP_CONN_STATE_CONNECTING:
+    case UCT_TCP_EP_CONN_STATE_WAITING_ACK:
+        if (old_conn_state == UCT_TCP_EP_CONN_STATE_CLOSED) {
+            uct_tcp_iface_outstanding_inc(iface);
+        }
+        break;
+    case UCT_TCP_EP_CONN_STATE_ACCEPTING:
+        break;
+    case UCT_TCP_EP_CONN_STATE_CONNECTED:
+        if (old_conn_state == UCT_TCP_EP_CONN_STATE_WAITING_ACK) {
```

Comment:
can we add assertions that old_conn state was either accepting or waiting_ack? relevant for other cases